### ISSUE TITLE
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-* fail when NINJA is set in calling env

### ISSUE THREAD

**Issue by mgorny** [mesonbuild/meson-python#314]
I've noticed the following tests failing because I have `NINJA=samu` set in my environment:

```
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-patchelf-pure] FAILED                                          [ 30%]
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-patchelf-library] FAILED                                       [ 31%]
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-nopatchelf-pure] FAILED                                        [ 33%]
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-nopatchelf-library] FAILED                                     [ 34%]
```

To reproduce:

```
export NINJA=samu
nox -e test-3.9
```

I suppose the test suite could either mock or unset it.

<details><summary>Full output</summary>

```pytb
nox > Running session test-3.9
nox > Re-using existing virtual environment at .nox/test-3-9.
nox > python -m pip install '.[test]'
nox > pytest --showlocals -vv --cov --cov-report=html:/tmp/meson-python/.nox/test-3-9/htmlcov --cov-report=xml:/tmp/meson-python/.nox/test-3-9/coverage-3.9.xml
========================================================= test session starts =========================================================
platform linux -- Python 3.9.16, pytest-7.2.1, pluggy-1.0.0 -- /tmp/meson-python/.nox/test-3-9/bin/python
cachedir: .pytest_cache
rootdir: /tmp/meson-python, configfile: pyproject.toml, testpaths: tests
plugins: cov-4.0.0, mock-3.10.0
collected 69 items                                                                                                                    

tests/test_consistency.py::test_pyproject_dependencies PASSED                                                                   [  1%]
tests/test_examples.py::test_spam PASSED                                                                                        [  2%]
tests/test_metadata.py::test_no_pep621 PASSED                                                                                   [  4%]
tests/test_metadata.py::test_pep621 PASSED                                                                                      [  5%]
tests/test_metadata.py::test_dynamic_version PASSED                                                                             [  7%]
tests/test_output.py::test_colors[False-env0-False] PASSED                                                                      [  8%]
tests/test_output.py::test_colors[True-env1-True] PASSED                                                                        [ 10%]
tests/test_output.py::test_colors[False-env2-False] PASSED                                                                      [ 11%]
tests/test_output.py::test_colors[True-env3-False] PASSED                                                                       [ 13%]
tests/test_output.py::test_colors[False-env4-True] PASSED                                                                       [ 14%]
tests/test_output.py::test_colors[True-env5-True] PASSED                                                                        [ 15%]
tests/test_output.py::test_colors_conflict PASSED                                                                               [ 17%]
tests/test_pep517.py::test_get_requires_for_build_wheel[noninja-patchelf-pure] PASSED                                           [ 18%]
tests/test_pep517.py::test_get_requires_for_build_wheel[noninja-patchelf-library] PASSED                                        [ 20%]
tests/test_pep517.py::test_get_requires_for_build_wheel[noninja-nopatchelf-pure] PASSED                                         [ 21%]
tests/test_pep517.py::test_get_requires_for_build_wheel[noninja-nopatchelf-library] PASSED                                      [ 23%]
tests/test_pep517.py::test_get_requires_for_build_wheel[oldninja-patchelf-pure] PASSED                                          [ 24%]
tests/test_pep517.py::test_get_requires_for_build_wheel[oldninja-patchelf-library] PASSED                                       [ 26%]
tests/test_pep517.py::test_get_requires_for_build_wheel[oldninja-nopatchelf-pure] PASSED                                        [ 27%]
tests/test_pep517.py::test_get_requires_for_build_wheel[oldninja-nopatchelf-library] PASSED                                     [ 28%]
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-patchelf-pure] FAILED                                          [ 30%]
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-patchelf-library] FAILED                                       [ 31%]
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-nopatchelf-pure] FAILED                                        [ 33%]
tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-nopatchelf-library] FAILED                                     [ 34%]
tests/test_pep517.py::test_invalid_config_settings PASSED                                                                       [ 36%]
tests/test_pep517.py::test_invalid_config_settings_suggest PASSED                                                               [ 37%]
tests/test_pep518.py::test_pep518[sdist_to_wheel-purelib-and-platlib] PASSED                                                    [ 39%]
tests/test_pep518.py::test_pep518[wheel_directly-purelib-and-platlib] PASSED                                                    [ 40%]
tests/test_project.py::test_name[library] PASSED                                                                                [ 42%]
tests/test_project.py::test_name[library-pep621] PASSED                                                                         [ 43%]
tests/test_project.py::test_version[library] PASSED                                                                             [ 44%]
tests/test_project.py::test_version[library-pep621] PASSED                                                                      [ 46%]
tests/test_project.py::test_unsupported_dynamic PASSED                                                                          [ 47%]
tests/test_project.py::test_unsupported_python_version PASSED                                                                   [ 49%]
tests/test_project.py::test_user_args PASSED                                                                                    [ 50%]
tests/test_project.py::test_unknown_user_args[top-level] PASSED                                                                 [ 52%]
tests/test_project.py::test_unknown_user_args[meson-args] PASSED                                                                [ 53%]
tests/test_project.py::test_install_tags PASSED                                                                                 [ 55%]
tests/test_sdist.py::test_contents PASSED                                                                                       [ 56%]
tests/test_sdist.py::test_contents_subdirs PASSED                                                                               [ 57%]
tests/test_sdist.py::test_contents_unstaged PASSED                                                                              [ 59%]
tests/test_sdist.py::test_executable_bit PASSED                                                                                 [ 60%]
tests/test_sdist.py::test_generated_files PASSED                                                                                [ 62%]
tests/test_tags.py::test_wheel_tag PASSED                                                                                       [ 63%]
tests/test_tags.py::test_macos_platform_tag SKIPPED (macOS specific test)                                                       [ 65%]
tests/test_tags.py::test_python_host_platform SKIPPED (macOS specific test)                                                     [ 66%]
tests/test_tags.py::test_tag_empty_wheel PASSED                                                                                 [ 68%]
tests/test_tags.py::test_tag_purelib_wheel PASSED                                                                               [ 69%]
tests/test_tags.py::test_tag_platlib_wheel PASSED                                                                               [ 71%]
tests/test_tags.py::test_tag_stable_abi PASSED                                                                                  [ 72%]
tests/test_tags.py::test_tag_mixed_abi PASSED                                                                                   [ 73%]
tests/test_wheel.py::test_scipy_like PASSED                                                                                     [ 75%]
tests/test_wheel.py::test_contents PASSED                                                                                       [ 76%]
tests/test_wheel.py::test_purelib_and_platlib PASSED                                                                            [ 78%]
tests/test_wheel.py::test_pure PASSED                                                                                           [ 79%]
tests/test_wheel.py::test_configure_data PASSED                                                                                 [ 81%]
tests/test_wheel.py::test_local_lib PASSED                                                                                      [ 82%]
tests/test_wheel.py::test_contents_license_file PASSED                                                                          [ 84%]
tests/test_wheel.py::test_executable_bit PASSED                                                                                 [ 85%]
tests/test_wheel.py::test_detect_wheel_tag_module PASSED                                                                        [ 86%]
tests/test_wheel.py::test_detect_wheel_tag_script PASSED                                                                        [ 88%]
tests/test_wheel.py::test_rpath PASSED                                                                                          [ 89%]
tests/test_wheel.py::test_uneeded_rpath PASSED                                                                                  [ 91%]
tests/test_wheel.py::test_entrypoints PASSED                                                                                    [ 92%]
tests/test_wheel.py::test_top_level_modules PASSED                                                                              [ 94%]
tests/test_wheel.py::test_editable PASSED                                                                                       [ 95%]
tests/test_wheel.py::test_editable_broken_non_existent_build_dir PASSED                                                         [ 97%]
tests/test_wheelfile.py::test_basic PASSED                                                                                      [ 98%]
tests/test_wheelfile.py::test_source_date_epoch PASSED                                                                          [100%]/tmp/meson-python/.nox/test-3-9/lib/python3.9/site-packages/coverage/html.py:111: CoverageWarning: No contexts were measured
  self.coverage._warn("No contexts were measured")


============================================================== FAILURES ===============================================================
______________________________________ test_get_requires_for_build_wheel[newninja-patchelf-pure] ______________________________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac08456880>, package = 'pure', system_patchelf = 'patchelf'
ninja = '1.8.3'

    @pytest.mark.parametrize('package', ['pure', 'library'])
    @pytest.mark.parametrize('system_patchelf', ['patchelf', None], ids=['patchelf', 'nopatchelf'])
    @pytest.mark.parametrize('ninja', [None, '1.8.1', '1.8.3'], ids=['noninja', 'oldninja', 'newninja'])
    def test_get_requires_for_build_wheel(monkeypatch, package, system_patchelf, ninja):
        def which(prog: str) -> bool:
            if prog == 'patchelf':
                return system_patchelf
            if prog == 'ninja':
                return ninja and 'ninja'
            if prog in ('ninja-build', 'samu'):
                return None
            # smoke check for the future if we add another usage
            raise AssertionError(f'Called with {prog}, tests not expecting that usage')
    
        subprocess_run = subprocess.run
    
        def run(cmd: List[str], *args: object, **kwargs: object) -> subprocess.CompletedProcess:
            if cmd == ['ninja', '--version']:
                return subprocess.CompletedProcess(cmd, 0, f'{ninja}\n', '')
            return subprocess_run(cmd, *args, **kwargs)
    
        monkeypatch.setattr(shutil, 'which', which)
        monkeypatch.setattr(subprocess, 'run', run)
    
        expected = set()
    
        ninja_available = ninja is not None and [int(x) for x in ninja.split('.')] >= [1, 8, 2]
    
        if not ninja_available:
            expected |= {mesonpy._depstr.ninja}
    
        if (
            system_patchelf is None and sys.platform.startswith('linux')
            and (not ninja_available or (ninja_available and package != 'pure'))
        ):
            expected |= {mesonpy._depstr.patchelf}
    
        with chdir(package_dir / package):
>           assert set(mesonpy.get_requires_for_build_wheel()) == expected
E           AssertionError: assert {'ninja >= 1.8.2'} == set()
E             Extra items in the left set:
E             'ninja >= 1.8.2'
E             Full diff:
E             - set()
E             + {'ninja >= 1.8.2'}

expected   = set()
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac08456880>
ninja      = '1.8.3'
ninja_available = True
package    = 'pure'
run        = <function test_get_requires_for_build_wheel.<locals>.run at 0x7fac0845fe50>
subprocess_run = <function run at 0x7fac0cc8e8b0>
system_patchelf = 'patchelf'
which      = <function test_get_requires_for_build_wheel.<locals>.which at 0x7fac0846bca0>

tests/test_pep517.py:58: AssertionError
____________________________________ test_get_requires_for_build_wheel[newninja-patchelf-library] _____________________________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac08589eb0>, package = 'library', system_patchelf = 'patchelf'
ninja = '1.8.3'

    @pytest.mark.parametrize('package', ['pure', 'library'])
    @pytest.mark.parametrize('system_patchelf', ['patchelf', None], ids=['patchelf', 'nopatchelf'])
    @pytest.mark.parametrize('ninja', [None, '1.8.1', '1.8.3'], ids=['noninja', 'oldninja', 'newninja'])
    def test_get_requires_for_build_wheel(monkeypatch, package, system_patchelf, ninja):
        def which(prog: str) -> bool:
            if prog == 'patchelf':
                return system_patchelf
            if prog == 'ninja':
                return ninja and 'ninja'
            if prog in ('ninja-build', 'samu'):
                return None
            # smoke check for the future if we add another usage
            raise AssertionError(f'Called with {prog}, tests not expecting that usage')
    
        subprocess_run = subprocess.run
    
        def run(cmd: List[str], *args: object, **kwargs: object) -> subprocess.CompletedProcess:
            if cmd == ['ninja', '--version']:
                return subprocess.CompletedProcess(cmd, 0, f'{ninja}\n', '')
            return subprocess_run(cmd, *args, **kwargs)
    
        monkeypatch.setattr(shutil, 'which', which)
        monkeypatch.setattr(subprocess, 'run', run)
    
        expected = set()
    
        ninja_available = ninja is not None and [int(x) for x in ninja.split('.')] >= [1, 8, 2]
    
        if not ninja_available:
            expected |= {mesonpy._depstr.ninja}
    
        if (
            system_patchelf is None and sys.platform.startswith('linux')
            and (not ninja_available or (ninja_available and package != 'pure'))
        ):
            expected |= {mesonpy._depstr.patchelf}
    
        with chdir(package_dir / package):
>           assert set(mesonpy.get_requires_for_build_wheel()) == expected
E           AssertionError: assert {'ninja >= 1.8.2'} == set()
E             Extra items in the left set:
E             'ninja >= 1.8.2'
E             Full diff:
E             - set()
E             + {'ninja >= 1.8.2'}

expected   = set()
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac08589eb0>
ninja      = '1.8.3'
ninja_available = True
package    = 'library'
run        = <function test_get_requires_for_build_wheel.<locals>.run at 0x7fac0849f670>
subprocess_run = <function run at 0x7fac0cc8e8b0>
system_patchelf = 'patchelf'
which      = <function test_get_requires_for_build_wheel.<locals>.which at 0x7fac0847e040>

tests/test_pep517.py:58: AssertionError
_____________________________________ test_get_requires_for_build_wheel[newninja-nopatchelf-pure] _____________________________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac084412b0>, package = 'pure', system_patchelf = None, ninja = '1.8.3'

    @pytest.mark.parametrize('package', ['pure', 'library'])
    @pytest.mark.parametrize('system_patchelf', ['patchelf', None], ids=['patchelf', 'nopatchelf'])
    @pytest.mark.parametrize('ninja', [None, '1.8.1', '1.8.3'], ids=['noninja', 'oldninja', 'newninja'])
    def test_get_requires_for_build_wheel(monkeypatch, package, system_patchelf, ninja):
        def which(prog: str) -> bool:
            if prog == 'patchelf':
                return system_patchelf
            if prog == 'ninja':
                return ninja and 'ninja'
            if prog in ('ninja-build', 'samu'):
                return None
            # smoke check for the future if we add another usage
            raise AssertionError(f'Called with {prog}, tests not expecting that usage')
    
        subprocess_run = subprocess.run
    
        def run(cmd: List[str], *args: object, **kwargs: object) -> subprocess.CompletedProcess:
            if cmd == ['ninja', '--version']:
                return subprocess.CompletedProcess(cmd, 0, f'{ninja}\n', '')
            return subprocess_run(cmd, *args, **kwargs)
    
        monkeypatch.setattr(shutil, 'which', which)
        monkeypatch.setattr(subprocess, 'run', run)
    
        expected = set()
    
        ninja_available = ninja is not None and [int(x) for x in ninja.split('.')] >= [1, 8, 2]
    
        if not ninja_available:
            expected |= {mesonpy._depstr.ninja}
    
        if (
            system_patchelf is None and sys.platform.startswith('linux')
            and (not ninja_available or (ninja_available and package != 'pure'))
        ):
            expected |= {mesonpy._depstr.patchelf}
    
        with chdir(package_dir / package):
>           assert set(mesonpy.get_requires_for_build_wheel()) == expected
E           AssertionError: assert {'ninja >= 1.8.2', 'patchelf >= 0.11.0'} == set()
E             Extra items in the left set:
E             'ninja >= 1.8.2'
E             'patchelf >= 0.11.0'
E             Full diff:
E             - set()
E             + {'ninja >= 1.8.2', 'patchelf >= 0.11.0'}

expected   = set()
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac084412b0>
ninja      = '1.8.3'
ninja_available = True
package    = 'pure'
run        = <function test_get_requires_for_build_wheel.<locals>.run at 0x7fac0849f040>
subprocess_run = <function run at 0x7fac0cc8e8b0>
system_patchelf = None
which      = <function test_get_requires_for_build_wheel.<locals>.which at 0x7fac0846bb80>

tests/test_pep517.py:58: AssertionError
___________________________________ test_get_requires_for_build_wheel[newninja-nopatchelf-library] ____________________________________

monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac0859e730>, package = 'library', system_patchelf = None, ninja = '1.8.3'

    @pytest.mark.parametrize('package', ['pure', 'library'])
    @pytest.mark.parametrize('system_patchelf', ['patchelf', None], ids=['patchelf', 'nopatchelf'])
    @pytest.mark.parametrize('ninja', [None, '1.8.1', '1.8.3'], ids=['noninja', 'oldninja', 'newninja'])
    def test_get_requires_for_build_wheel(monkeypatch, package, system_patchelf, ninja):
        def which(prog: str) -> bool:
            if prog == 'patchelf':
                return system_patchelf
            if prog == 'ninja':
                return ninja and 'ninja'
            if prog in ('ninja-build', 'samu'):
                return None
            # smoke check for the future if we add another usage
            raise AssertionError(f'Called with {prog}, tests not expecting that usage')
    
        subprocess_run = subprocess.run
    
        def run(cmd: List[str], *args: object, **kwargs: object) -> subprocess.CompletedProcess:
            if cmd == ['ninja', '--version']:
                return subprocess.CompletedProcess(cmd, 0, f'{ninja}\n', '')
            return subprocess_run(cmd, *args, **kwargs)
    
        monkeypatch.setattr(shutil, 'which', which)
        monkeypatch.setattr(subprocess, 'run', run)
    
        expected = set()
    
        ninja_available = ninja is not None and [int(x) for x in ninja.split('.')] >= [1, 8, 2]
    
        if not ninja_available:
            expected |= {mesonpy._depstr.ninja}
    
        if (
            system_patchelf is None and sys.platform.startswith('linux')
            and (not ninja_available or (ninja_available and package != 'pure'))
        ):
            expected |= {mesonpy._depstr.patchelf}
    
        with chdir(package_dir / package):
>           assert set(mesonpy.get_requires_for_build_wheel()) == expected
E           AssertionError: assert {'ninja >= 1.8.2', 'patchelf >= 0.11.0'} == {'patchelf >= 0.11.0'}
E             Extra items in the left set:
E             'ninja >= 1.8.2'
E             Full diff:
E             - {'patchelf >= 0.11.0'}
E             + {'ninja >= 1.8.2', 'patchelf >= 0.11.0'}

expected   = {'patchelf >= 0.11.0'}
monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7fac0859e730>
ninja      = '1.8.3'
ninja_available = True
package    = 'library'
run        = <function test_get_requires_for_build_wheel.<locals>.run at 0x7fac0848fa60>
subprocess_run = <function run at 0x7fac0cc8e8b0>
system_patchelf = None
which      = <function test_get_requires_for_build_wheel.<locals>.which at 0x7fac0849f160>

tests/test_pep517.py:58: AssertionError

---------- coverage: platform linux, python 3.9.16-final-0 -----------
Coverage HTML written to dir /tmp/meson-python/.nox/test-3-9/htmlcov
Coverage XML written to file /tmp/meson-python/.nox/test-3-9/coverage-3.9.xml

======================================================= short test summary info =======================================================
SKIPPED [1] tests/test_tags.py:36: macOS specific test
SKIPPED [1] tests/test_tags.py:47: macOS specific test
FAILED tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-patchelf-pure] - AssertionError: assert {'ninja >= 1.8.2'} == set()
FAILED tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-patchelf-library] - AssertionError: assert {'ninja >= 1.8.2'} == set()
FAILED tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-nopatchelf-pure] - AssertionError: assert {'ninja >= 1.8.2', 'patchelf >= 0.11.0'} == set()
FAILED tests/test_pep517.py::test_get_requires_for_build_wheel[newninja-nopatchelf-library] - AssertionError: assert {'ninja >= 1.8.2', 'patchelf >= 0.11.0'} == {'patchelf >= 0.11.0'}
========================================= 4 failed, 63 passed, 2 skipped in 102.95s (0:01:42) =========================================
nox > Command pytest --showlocals -vv --cov --cov-report=html:/tmp/meson-python/.nox/test-3-9/htmlcov --cov-report=xml:/tmp/meson-python/.nox/test-3-9/coverage-3.9.xml failed with exit code 1
nox > Session test-3.9 failed.

```
</details>

**Comment by dnicolodi** [comment]
I have a quick fix in #315. However, I think something is off with the code that verifies the `ninja` version when the `NINJA` environment variable is set: if the variable points to a path that does not exist or to an executable that reports a too old version, we ask the `ninja` package to be installed, however we don't reset the `NINJA` environment variable, thus Meson will fail when meson-python executes `meson setup`.

**Comment by mgorny** [comment]
Thank you! Cool if my report might have helped you find another problem ;-).
